lua/ttt2/libraries/spawn.lua

```diff
@@ -0,0 +1,303 @@
+	local spawnPoints = spawn.GetSpawnPointsAroundSpawn(ply, unsafePos)
+
+	for i = 1, #spawnPoints do
+		local spawnPoint = spawnPoints[i]
+
+		if not spawn.IsSpawnPointSafe(ply, spawnPoint, false) then continue end
+
+		return spawnPoint
+	end
+end
+
+---
+-- Returns a list of all spawn entities found on the map that are valid to spawn players.
+-- @warning Don't use 'info_player_start' unless absolutely necessary, because eg. TF2 uses
+-- it for observer starts that are in places where players cannot really spawn well.
+-- Therefore by default 'info_player_start' is not included in the search. However if 'forceAll'
+-- is set to true or no other map spawn was found, these spawn entities will be included.
+-- @param boolean forceAll Shouldn't used unless absolutely necessary because it includes 'info_player_start' spawns
+-- @return table Returns a table of unsafe spawn entities
+-- @realm server
+function spawn.GetPlayerSpawnEntities(forceAll)
+	local tbl = {}
+
+	for i = 1, #spawnTypes do
+		local classname = spawnTypes[i]
+		local entsTbl = entsFindByClass(classname)
+
+		for k = 1, #entsTbl do
+			if entsTbl[k].markAsRemoved then continue end
+
+			tbl[#tbl + 1] = entsTbl[k]
+		end
+	end
+
+	-- Don't use info_player_start unless absolutely necessary, because eg. TF2
+	-- uses it for observer starts that are in places where players cannot really
+	-- spawn well. At all.
+	if forceAll or #tbl == 0 then
+		local startTbl = entsFindByClass("info_player_start")
+
+		for k = 1, #startTbl do
+			if startTbl[k].markAsRemoved then continue end
+
+			tbl[#tbl + 1] = startTbl[k]
+		end
+	end
+
+	return tbl
+end
+
+---
+-- Gets a table of spawnpoints on the map.
+-- @note These spawn points are not shuffled.
+-- @return table Returns a table of spawnpoints as @{Vector}s
+-- @realm server
+function spawn.GetPlayerSpawnPointTable()
+	local spawnEnts = spawn.GetPlayerSpawnEntities(false)
+	local spawnPoints = {}
+
+	for i = 1, #spawnEnts do
+		spawnPoints[i] = spawnEnts[i]:GetPos()
+	end
+
+	return spawnPoints
+end
+
+---
+-- Gets a safe random player spawn entity. If no free and safe spawnpoint is found, it tries
+-- to create its own by searching around existing ones.
+-- @param Player ply The player that should receive their spawn point
+-- @return Entity A safe spawn entity
+-- @realm server
+function spawn.GetRandomPlayerSpawnEntity(ply)
+	-- One might think that we have to regenerate our spawnpoint
+	-- cache. Otherwise, any riggedSpawnPoints spawn entities would not get reused, and
+	-- MORE new entities would be made instead. In reality, the map cleanup at
+	-- round start will remove our riggedSpawnPoints spawns, and we'll have to create new
+	-- ones anyway.
+	if #spawn.cachedSpawnEntities == 0 or not IsTableOfEntitiesValid(spawn.cachedSpawnEntities) then
+		spawn.cachedSpawnEntities = spawn.GetPlayerSpawnEntities(false)
+	end
+
+	if #spawn.cachedSpawnEntities == 0 then
+		Error("No spawn entity found!\n")
```

Comment:
Are you sure using `Error` instead of `ErrorNoHalt` ?

Reply:
I have no idea. It has been this way in the original TTT code. Probably an error is ok here if no player is able to spawn.